./tools/check-typo-since trunk

gasche opened the pull request:
I have not learned to use the check-typo pre-commit-hook yet (and I think there are various situations where I want to run check-typo at other times than at pre-commits), but I already find just running `./tools/check-typo` (to check the whole repository) very slow, it takes slightly above one minute on my machine.

This PR proposes an auxiliary tool, `check-typo-since`, which takes a git reference as argument, and runs `check-typo` on all the files that have changed since this reference -- due to further commit, or not-yet-committed changes. The following could be useful

```
./tools/check-typo-since trunk  # check my whole feature branch
./tools/check-typo-since HEAD~1 # check last commit
./tools/check-typo-since HEAD # check uncommitted changes
```

I first considered adding this feature to `check-typo` directly (`--since trunk`), but the control flow of `check-typo` is horrible spaghetti and I don't want to touch it.

To make `check-typo-since` work well, I had to make some tweaks to `.gitattributes`. In particular, instead of having `/manual ocaml-typo=prune`, which tells that the `manual` directory is to be pruned, I changed it into `/manual** ocaml-typo=prune`, which tells that `manual` and any file underneath is to be pruned. This makes the implementation of `check-typo-since` simple and easy, but one has to remember to always use the child-closure operator `**` when pruning in gitattributes.

(I realize now that some of the logic is very close to the code of the pre-commit-hook, but that I made different choices which could also be used to simplify the pre-commit-hook's logic, if @dra27 agrees that they are sensible choices.)

gasche committed:
add a variant of check-typo to compare against a git reference

./tools/check-typo-since trunk
./tools/check-typo-since HEAD~10

In most cases, this should be much faster than running check-typo on
the whole directory.

gasche committed:
.gitattributes: prune directories recursively

Before this change, check-typo would run on manual/Makefile for
example, while this file lives within a pruned directory so it
ought to be ignored by the tool.

Note: the check-typo code seems to assume that the only pruned things
are directory, it prints "pruned directory ..." when something is
pruned. I haven't changed this part of the logic; but note that normal
./check-typo invocation will only check pruning for directories.

dra27 commented on .gitattributes:
```diff
@@ -52,20 +52,22 @@ README*                  ocaml-typo=missing-header
 asmcomp/power/NOTES.md   ocaml-typo=missing-header
 
-/config/gnu              ocaml-typo=prune
+/config/gnu**            ocaml-typo=prune
```
This (and the other two changes) are slightly in conflict with the reasoning behind https://github.com/ocaml/ocaml/pull/1148/commits/2adc6af92a60815f76260928ae2829336bf2ade3 - not necessarily a problem. The alternative is to steal the logic from https://github.com/ocaml/ocaml/blob/trunk/tools/ci/travis/travis-ci.sh#L187-L203

gasche replied:
I don't like the logic from travis-ci.sh; I don't think that the consumers of `check-typo` should each reproduce logic that is easily expressed in it or in gitattributes. (We could have the recurse-on-dirname logic in check-typo itself, but it would probably not help with the slowness of it).

The reasoning is "have the attributes apply to exactly the right files, no more". I would say that taking the transitive closure of a parent directory is not really a loss of precision, but a different choice of semantics. On the other hand, it may be that with the current formulation `/config/gnufoo` would also be concerned; I can test and, if this is the case, change it to have both `/config/gnu` and `/config/gnu/**`.

On the other hand, one option is to say "I want to be able to run check-typo on manual files, when running the tool from the manual directory itself for example, just they shouldn't be traversed on runs from the root". Then the change would be clearly wrong. But I don't think that we want that -- either we decide to keep a file under strict discipline, or we should give up checking on it all the time, I don't see a meaningful compromise in-between.

dra27 commented on .gitattributes:
```diff
@@ -52,20 +52,22 @@ README*                  ocaml-typo=missing-header
 /tools/ci/appveyor/appveyor_build.sh       ocaml-typo=non-ascii
 
+# depend files are auto-generated
+.depend                  ocaml-typo=missing-header
```
This is supposed to be an automatic rule, but I think they're only invoked when check-typo is running in it's original "entire repo" mode. I think it would be better to fix that, than paper over it here - or at least it should be that the comment notes that the line is only required because of a bug in check-typo.

gasche replied:
Thanks, I will investigate.

dra27 commented on .gitattributes:
```diff
@@ -52,20 +52,22 @@ README*                  ocaml-typo=missing-header
 /tools/ci/appveyor/appveyor_build.sh       ocaml-typo=non-ascii
 
+# depend files are auto-generated
+.depend                  ocaml-typo=missing-header
```
This is supposed to be an automatic rule, but I think they're only invoked when check-typo is running in it's original "entire repo" mode. I think it would be better to fix that, than paper over it here - or at least it should be that the comment notes that the line is only required because of a bug in check-typo.

dra27 commented on .gitattributes:
```diff
@@ -52,20 +52,22 @@ README*                  ocaml-typo=missing-header
 asmcomp/power/NOTES.md   ocaml-typo=missing-header
 
-/config/gnu              ocaml-typo=prune
+/config/gnu**            ocaml-typo=prune
```
This (and the other two changes) are slightly in conflict with the reasoning behind https://github.com/ocaml/ocaml/pull/1148/commits/2adc6af92a60815f76260928ae2829336bf2ade3 - not necessarily a problem. The alternative is to steal the logic from https://github.com/ocaml/ocaml/blob/trunk/tools/ci/travis/travis-ci.sh#L187-L203

dra27 commented on tools/check-typo-since:
```diff
@@ -0,0 +1,37 @@
+  do
+    $CHECK_TYPO --check-prune $f 2>/dev/null
+    CHECK_PRUNE=$?
+    if [[ -e $f && $CHECK_PRUNE > 0 ]]
```
I'm fairly sure `[[` is a bashism - though I'd change the shebang to be `#! /usr/bin/env bash` rather than remove the bashism.

gasche replied:
Thanks for the catch! I'll use the suggested shebang.

dra27 commented on tools/check-typo-since:
```diff
@@ -0,0 +1,37 @@
+
+check_typo_since() {
+  CHECK_TYPO=$(dirname $0)/check-typo
+  for f in $(git diff --name-only $1)
```
Filename processing should be done in a `while` loop - this will break weirdly if anyone commits a filename with a space. See the `CheckTypoTree` function in `travis-ci.sh` (although I discover that I forgot to quote `$path` in that function 😊)

gasche replied:
I'll look at this, thanks.

dra27 reviewed (commented):
Firstly, thanks for doing this, because I haven't got to writing it!

Given that this is a script for developers, I think it better to have a version which maybe wants more features in the future than spend ages without one.

It's worth fixing the couple of issues I've highlighted - longer term, doing `git diff` against `trunk` will cause the script to scan files which have been changed by upstream changes - it would be better (I think) to identify the merge-base with the branch specified and scan the commits inbetween for the list of files - which is what `travis-ci.sh` does.

gasche committed:
check-typo: make --check-prune faster

this comment special-cases the prune-detection logic to use the `git
check-attr` layer directly, instead of using the convenience function
`get_attrs ..` which parses its output.

On my machine, calling --check-prune on the testsuite files goes from
17s to 12s when this patch is applied.

gasche commented:
I have rebased this PR on top of #1910, and addressed the review comments -- but I think that another bashism-detection-check would not hurt.

gasche commented:
(Not sure why this was automatically closed, but I need to rebase it on trunk now that #1910 is in.)

gasche commented:
I just rebased this PR on top of trunk; reviews are warmly welcome. (But then again, this is not high priority.)

gasche commented:
Gentle ping: given that all issues of the previous review round have been addressed, could we accept this PR?

damiendoligez reviewed (commented):
Looks good to me, but I'll let @dra27 have another look.

gasche commented:
Ping :-)

gasche commented:
@damiendoligez (or @dra27) would you approve the PR? The feature is useful and the diff touches non-critical part of the codebase.

dra27 commented:
Sorry for the delay getting back to this one \<pages out m4sh and pages back in awk> ... LGTM!
